### DISCUSSION TITLE
InternalsVisibleTo to allow binding to internal view models

### DISCUSSION THREAD

**Discussion by andrewkdci** [unoplatform/uno#13657] [Q&A]
In our project, we are using Control Templates inside of a Resource Dictionary in order to template our controls. The Binding expression in order to bind to the view model of the control looks something like this:

`{Binding RelativeSource={RelativeSource TemplatedParent}, Path=ViewModel.MyValue}`

In order for this work, the control has a ViewModel property that is just a cast of the DataContext:

`internal MyViewModel ViewModel => DataContext as MyViewModel;`

(_Note that the view model is internal. This is intentional and desirable as part of (but only part of) our security paradigm._)

(_Also note that this pattern is a holdover from when this was a vanilla UWP app and we were using x:Bind and Resource Dictionaries with code behind. So please, no comments on whether this pattern is suboptimal or not. The purpose of this question is how to migrate the UWP app over to Uno as quickly and seamlessly as possible._)

When I run the app, I am getting a binding error that the ViewModel property cannot be found. I am 99% sure this is because the view model is internal.

I am thinking a quick fix for this would be to allow Uno to see internal properties by utilizing the `InternalsVisibleTo` assembly attribute. **My question is: What Uno assembly should I give internal visibility access to in order to allow data binding to internal properties**? This is even more confusing because my Windows head project only has a NuGet package reference to Uno.Core.Extensions.Logging.Singleton, which doesn't seem like the assembly I should target.

**Comment by andrewkdci** [comment]
It's dawning on me that this is perhaps a WinUI thing, at least for my Windows head. I will, however, need to do this for my WASM head too, which I believe **will** be an Uno assembly...

**Reply by jeromelaban** [reply]
Thanks for the question. If the issue you're getting is indeed in the Windows head, there's nothing much can do on Uno's end, aside from using `InternalsVisibleTo`, though even that may not be useful as it generally only controls compiler's visibility on members and types, not the binding engine's discovery process.

Could you provide a screenshot or copy of the error message as well as the platform you're getting it on?

**Comment by andrewkdci** [comment]
Thanks for the reply, @jeromelaban.

This is the error I get when I try to bind to an internal View Model for my Windows head:

`Error: BindingExpression path error: 'ViewModel' property not found on 'UnoViewModelUpdatesFromTemplate.MyControl'. BindingExpression: Path='ViewModel.MyValue' DataItem='UnoViewModelUpdatesFromTemplate.MyControl'; target element is 'Microsoft.UI.Xaml.Controls.TextBlock' (Name='null'); target property is 'Text' (type 'String')`

And this is the error I get when running my WASM head:

`fail: Uno.UI.DataBinding.BindingPropertyHelper[0]
      The [ViewModel] property getter does not exist on type [UnoViewModelUpdatesFromTemplate.MyControl]`

The workaround is that I change the "ViewModel" in my binding path to "DataContext": 

`{Binding RelativeSource={RelativeSource TemplatedParent}, Path=DataContext.MyValue}`

This works as long as the **property** is public, even if the actual object class is internal. For example:

```
    internal class MyViewModel : ViewModelBase
    {
        public int MyValue
        {
            get => _myValue;
            private set => SetProperty(ref _myValue, value);
        }
        private int _myValue;

        public MyViewModel NestedViewModel { get; set; }

        public MyViewModel(MyViewModel nestedViewModel)
        {
            NestedViewModel = nestedViewModel;
        }
    }
```

This binding does in fact still work:

`{Binding RelativeSource={RelativeSource TemplatedParent}, Path=DataContext.NestedViewModel.MyValue}`

ALSO, a couple more things to note on this question:

1. Just to be clear, this is not a bug or a feature request. I don't believe anything is broken. I'm just wondering what assembly I should give InternalsVisibleTo in order to change the "DataContext" to "ViewModel" in my binding, if this is even possible.
2. You're correct that there's not much that can be done or discussed on the Uno side for the Windows head, since the binding engine resides in the original WinUI framework when running that head. However I also need this to work for my WebAssembly head. Since Uno is handling the bindings for WASM, then I assume that it will be an Uno assembly that I need to give InternalsVisibleTo. That is really the main focus of this question, then.

Thanks!

**Reply by jeromelaban** [reply]
Thanks. If you note a behavior difference in the way Uno handles member accessibility, can you open a new issue with a small repro sample that demonstrates the difference?

**Reply by andrewkdci** [reply]
@jeromelaban Sure, will do.
